### DISCUSSION TITLE
New collection Review: netapp.um_info

### DISCUSSION THREAD

**Discussion by carchi8py** [ansible-collections/ansible-inclusion#18] [Resolved reviews]
We've created a few new collections for our other products (same team that currently support the existing NetApp collections). We do monthly update to each collection (And bug fixes through out the month if needed)

Ansible Galaxy: https://galaxy.ansible.com/netapp/um_info
Github: https://github.com/ansible-collections/netapp
Issues Tracker: https://github.com/ansible-collections/netapp/issues
Github Handles: @carchi8py @lonico

**Comment by jillr** [comment]
# Ansible Collections Checklist (short version)

_For details about the following points, refer to the [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst)._

Every comment should say whether the reviewer expects it to be addressed, or whether it's optional.

**Public availability and communication:**
- [x] published on Ansible Galaxy [Ansible Galaxy](https://galaxy.ansible.com) with version 1.0.0 or later
- [x] has a policy of releasing, versioning and deprecation announced to contributors and users in some way
  * Found: https://github.com/ansible-collections/netapp/issues/93
- [x] have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
- [x] have a publicly available issue tracker that does not require a paid level of service to create an account or view issues
- [x] have a public SCM repository, releases of the collection are tagged in this repository

**Standards and documentation:**
- [x] adheres to [semantic versioning](https://semver.org/)
- [x] follows licensing rules
  * Comment: Found GPL v2 in galaxy.yml, MUST be GPL v3 or later https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#licensing. There's GPL v3 in the modules
- [x] follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
- [x] follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html)
- [x] only uses allowed plugin types in the `plugins/` directory
- [x] has `README.md`
- [x] documentation, examples, and return use FQCNs for `M(..)`, examples, and `seealso` subsections
- [x] modules (or plugins) from ansible-base use `ansible.builtin.` as a FQCN prefix
- [x] documentation and return sections use `version_added:` containing the *collection* version for which an option, module or plugin was added
- [x] FQCNs is used for `extends_documentation_fragment:`, unless you are referring to doc_fragments from ansible-base

**Collection management:**
- [x] `galaxy.yml` having `tags` field set
- [x] collection dependencies must have a lower bound on the version which is at least 1.0.0, and are all part of the `ansible` package
- [x] `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
- [x] has changelog, preferably with `changelogs/changelog.yaml`

**Tests:**
- [x] passed `ansible-test sanity`
- [x] if `test/sanity/ignore*.txt` exists, it does not contain error codes listed [here](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#ci-testing)
- [x] has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
  * The repo README.md states the included collections support >=2.9 (should probably be 2.9.10).  Only found 2.10 testing in the [GH Action](https://github.com/ansible-collections/netapp/blob/master/.github/workflows/main.yml).
- [x] all CI tests run against every pull request
  * Sanity tests are run, unit tests are commented out. Recommended to be fixed.
- [x] all CI tests run against a commit that releases the collection; if they don't pass, the collection won't be released
- [x] all CI tests run regularly (nightly, or at least once per week)

**Reply by jillr** [reply]
There is already comments and discussion of the open items on the reviews for other netapp collections contained in the same repo, so I won't repeat those same items here.  Those will apply to all of the submitted collections.

**Comment by tadeboro** [comment]
Just a reminder: collections that want to be part of the Ansible 4.0.0 package need to get approval by 2021-04-26 ([full schedule](https://docs.ansible.com/ansible/devel/roadmap/COLLECTIONS_4.html#release-schedule)). This inclusion request still has some open issues that need to get addressed by then.

**Reply by tadeboro** [reply]
**Last reminder for inclusion submission**

Collections that want to be part of the Ansible 4.0.0 package need to get approval by 2021-04-26, which is just one week away ([full schedule](https://docs.ansible.com/ansible/devel/roadmap/COLLECTIONS_4.html#release-schedule)). Make sure you submit your collection for the final review a few days before the deadline (reviews take a non-trivial amount of time) and that your collection does meet all criteria listed in the [requirements document](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst).

**Comment by tadeboro** [comment]
Thank you for your submission, @carchi8py.

The items in the checklist can be in one of the following states:

  * Items with a checkmark and no attached comment are OK and need no intervention from your side.
  * Items with a checkmark that contain a comment will not block the inclusion process, but by following the advice given in the comment, the collection's quality can be improved.
  * Items with no checkmark and no extra details are waiting for evaluation.
  * Items with no checkmark that contain a comment need your attention.

Feel free to ask any questions and provide feedback. Thanks in advance for your understanding.

_For details about the following points, refer to the [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst)._

**Public availability and communication:**
- [x] published on Ansible Galaxy [Ansible Galaxy](https://galaxy.ansible.com) with version 1.0.0 or later
- [x] has a policy of releasing, versioning and deprecation announced to contributors and users in some way
- [x] have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
- [x] have a publicly available issue tracker that does not require a paid level of service to create an account or view issues
- [x] have a public SCM repository, releases of the collection are tagged in this repository

**Standards and documentation:**
- [x] adheres to [semantic versioning](https://semver.org/)
- [x] follows [licensing rules](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#licensing)
- [x] follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
      **Comment:** The netapp documentation fragment is invalid: https://github.com/ansible-collections/netapp/blob/main/ansible_collections/netapp/um_info/plugins/doc_fragments/netapp.py#L12-L19, but is not used anywhere. This will not block the inclusion, but should be fixed/simplified in the future.
- [x] follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html)
- [x] supports Python 2.6 or greater and Python 3.5 or greater. If it does not, read the [full guidelines](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#python-compatibility) to see if you qualify for an exception and document the unsupported [Python versions](https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#ansible-and-python-3) in the collection ``README.md`` and in every module and plugin (or in doc fragments)
- [x] only uses the [allowed plugin types](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#modules-plugins) in the `plugins/` directory
- [x] has `README.md`
- [x] documentation, examples, and return sections use FQCNs for the `M(..)` [format macros](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation) when referring to modules
- [x] modules (or plugins) from ansible-core use `ansible.builtin.` as a FQCN prefix
- [x] documentation and return sections use `version_added:` containing the *collection* version for which an option, module or plugin was added
- [x] FQCNs are used for `extends_documentation_fragment:`, unless you are referring to doc_fragments from ansible-core

**Collection management:**
- [x] `galaxy.yml` having `tags` field set
- [x] collection dependencies must have a lower bound on the version which is at least 1.0.0, and are all part of the `ansible` package
- [x] `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
- [x] has changelog, preferably with `changelogs/changelog.yaml`

**Tests:**
- [x] passed `ansible-test sanity`
- [x] if `test/sanity/ignore*.txt` exists, it does not contain error codes listed [here](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#ci-testing)
- [x] has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
- [x] all CI tests run against every pull request
- [x] all CI tests run against a commit that releases the collection; if they don't pass, the collection won't be released
- [x] all CI tests run regularly (nightly, or at least once per week)

**Reply by tadeboro** [reply]
I did the initial review of the collection. Once a few minor issues are fixed, this collection is ready to be included in the community Ansible package.

**Reply by jillr** [reply]
@tadeboro Where did you find the Code of Conduct? (I may just be overlooking it)

**Reply by tadeboro** [reply]
Ugh, my bad. I saw it in https://github.com/ansible-collections/netapp/tree/main/ansible_collections/netapp/cloudmanager#code-of-conduct. Multiple collections is a single repo is killing me. I did the same error in the azure review. Thank you for keeping me honest!

**Reply by lonico** [reply]
Addressing CoC, FCQNs, and runtime.ml issues.

**Reply by lonico** [reply]
changes pushed to GitHub.

**Reply by tadeboro** [reply]
Thank you, @carchi8py, for your perseverance. Once the next release hits Ansible Galaxy, this collection is ready for inclusion as far as I am concerned. Congratulations!

**Reply by carchi8py** [reply]
@tadeboro the update with these changes are now in galaxy https://galaxy.ansible.com/netapp/um_info

**Comment by felixfontein** [comment]
- The docs fragment's `notes:` entry has `\\_`, which translates to `\_` in the documentation (same as https://github.com/ansible-collections/ansible-inclusion/discussions/16#discussioncomment-646772).
- All modules in this collection are named `na_um_list_<object>` and seem to be modules that list certain objects. Is there a reason you are not naming them `_info`, as querying-only modules in Ansible are usually named?
- A `notes:` saying `Supports C(check_mode).` would be nice for every module.

**Reply by carchi8py** [reply]
Is there a way to deprecate a module in collections like there was in Ansible 2.9?

We have a lot of users currently using this collection, we don't want to break the existing playbooks. Is there a way to deprecate a module like we did back in ansible 2.7, 2.8, 2.9 https://github.com/ansible/ansible/tree/stable-2.9/lib/ansible/modules/storage/netapp (The modules with _ as the first character in their names)

**Reply by jillr** [reply]
@carchi8py After you've moved the filename to start with the _ you can add a `deprecated:` section to the module docs and a `deprecation:` block to `meta/runtime.yml`.
https://docs.ansible.com/ansible/latest/dev_guide/module_lifecycle.html
https://docs.ansible.com/ansible/latest/dev_guide/developing_collections.html#meta-directory

**Reply by lonico** [reply]
@felixfontein Is it OK if we defer renaming the modules from xxx_list to xxx_info by a few weeks?   We'd like to do it as a 21.6.0 release, but we are also restructuring our repos.   Or do we have to do it today?

**Reply by felixfontein** [reply]
Please do **not** use the `_` prefix for deprecating modules, that does not work in collections. The documentation in https://docs.ansible.com/ansible/latest/dev_guide/module_lifecycle.html#deprecating-modules is for modules in the ansible/ansible repository and not for modules in collections. Parts of it also apply to collections (in particular the part on `deprecated:` and its contents), but neither the underscore nor `porting_guide_x.y.rst` do apply to collections. (In fact it mentiones a bit below that the documentation needs to be updated for collections, and also mentions `routing.yml` which was a long time ago renamed to `runtime.yml`.)

About renaming: please do **not** deprecate the modules, but rename them and deprecate the old name. For that, basically rename the module, create a symlink from the old name to the new name, and call `module.deprecate()` in case the module was called with its old name. Also the rename should be mentioned in the documentation (so that users aren't confused). The only reason to deprecate + add new module instead of renaming is when you want to break the module's API, which I think you do not want to do.

Finally: I would probably rename the modules, but wait a bit with deprecation of the old name. This gives users a bit of time where both names work without deprecation, which eases transition for users of your collection.

**Reply by jillr** [reply]
My mistake, thanks @felixfontein (it's quite really confusing though).

**Reply by felixfontein** [reply]
I created a PR to update that document here: https://github.com/ansible/ansible/pull/74413

**Comment by felixfontein** [comment]
This collection has been prepared for inclusion in Ansible 4 in this PR: https://github.com/ansible-community/ansible-build-data/pull/66

**Comment by tadeboro** [comment]
The `netapp.um_info` Ansible Collection was accepted into the Ansible community package. The relevant part of the community meeting starts [here](https://meetbot.fedoraproject.org/ansible-community/2021-04-28/ansible_community_meeting.2021-04-28-18.00.log.html#l-236).

Congratulations!

**Reply by felixfontein** [reply]
We do expect that the modules are renamed to `_info` in the next minor release(s) so that in Ansible 4.1.0 the new names will be available. 4.1.0 should be released on June 8th, so if you do this in the 21.6.0 release, everything should be fine (assuming 21.6.0 is released before June 8th).
